Refactor and cleanup treatment of keyspace IDs and KeyRange

jeremycole opened the pull request:
## Description

Fixes https://github.com/vitessio/vitess/issues/12535

Initially we encountered a bug where unevenly split shards with a set of key ranges as (... `000280-000300`, `0003-`) did not properly initialize shard `000280-000300` due to believing it overlapped with `0003-`. (This was due to a comparison where `000300` should have been equal to `0003` but was not.)

In the course of fixing this bug, I decided to refactor and cleanup the treatment of keyspace IDs and `topodatapb.KeyRange` values overall. The treatment of keyspace IDs was very inconsistent resulting in many comparisons not properly handling the `KeyRange`-specific edge cases (mainly due to logic using `bytes.Compare` scattered across the codebase):

* The `Start` and `End` fields comparing empty/zero values as minimum-key and maximum-key, respectively.
* Inconsistently handling `Start` or `End` fields containing `nil` and zero-length `[]byte`.
* Handling of functionally equal but differently represented values, primarily different-length keys e.g. `80` vs. `8000`. (Attempts were made previously to resolve this with `addPadding` which proved incomplete; my initial fix which we shipped/tested internally added calls to `addPadding` to everywhere keyspace IDs were being passed, which was effective but very ugly and poorly factored.)

In addition I've added a number of tests and improved test coverage of much of the existing code, as well as clarifying existing comments and adding new ones. All of the existing tests continue to pass, with one exception which appears to be due to a bug in the original code which was then codified in the corresponding test:

### Side quest: a bug fix for `TestKeyRangeContiguous`

`TestKeyRangeContiguous` tested `"-"` and `"-40"` and expected a result of `true`, but those key ranges are most definitely not contiguous. The corresponding code had the following set of logic:

```golang
	if left == nil {
		return right == nil || (len(right.Start) == 0 && len(right.End) == 0)
	}
	if right == nil {
		return len(left.Start) == 0 && len(left.End) == 0
	}
```

That's attempting to handle `nil` values (although inconsistently) but treating `nil` as a missing value (inconsistently), whereas the rest of the code treats `nil` as full-range. I believe this should as well. I've replaced it with new logic, which I believe to be more correct:

```golang
	if KeyRangeIsComplete(a) || KeyRangeIsComplete(b) {
		return false // no two KeyRange values can be contiguous if either is the complete range
	}
```

If there is some reason this should treat `nil` differently, let me know and I'll be happy to fix that and document it. However, there is only one call site and it doesn't read to me like it requires any special/different handling, so I believe this was just a misunderstanding.

### Future work

There's still a fair bit of work to be done to clean this up, I believe at least the following would be reasonable:

* Introduce a new explicit `type KeyspaceID` or similar for keyspace IDs and replace usage of `[]byte` to the extent possible.
  * Adjust the naming/namespacing and interfaces for all the various functions on them to accomodate.
  * Additionally remove the `type Uint64Key` (which is only used in tests) in favor of the new `type`.
* Perhaps introduce a first-class `type KeyRange`, re-organize all the various comparison functions, and consistently handle both parsing and formatting KeyRanges in a single set of functions.
* Remove `EvenShardsKeyRange` and `GenerateShardRanges` which behave very questionably at best and are rarely used.

I did not attempt to do any of these here to keep the PR size reasonable, but if there is some level of agreement (or at least a lack of disagreement) I'd be happy to undertake them.

## Related Issue(s)


## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

## Deployment Notes

There should not be any user-facing impact of this change, as it's purely backend and bug fixing existing functionality which was perhaps never used before.

harshit-gangal commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+// KeyRangesIntersect returns true if some part of KeyRange a and b overlap, meaning that some keyspace ID values
+// exist in both a and b.
+func KeyRangesIntersect(a, b *topodatapb.KeyRange) bool {
+	if a == nil || b == nil {
+		return true
 	}
-	return bytes.Equal(addPadding(left.Start), addPadding(right.Start))
+
+	return (Empty(a.End) || Less(b.Start, a.End)) && (Empty(b.End) || Less(a.Start, b.End))
 }
```
I think I got my answer, It is not.

harshit-gangal reviewed (approved):
This is a good change and I think it will resolve the issue related to key range matching.

mattlord commented on go/vt/key/key.go:
```diff
@@ -351,11 +351,63 @@ func ParseShardingSpec(spec string) ([]*topodatapb.KeyRange, error) {
+	// Determine the number of bytes which are required to represent any
+	// KeyRange start or end for the given n.
+	// This is required to trim the returned values to the same length e.g.
+	// (256, 512) should return 8000-8080 as shard key range.
```
Nice. Quick confirmation:
```
$ vtctldclient GenerateShardRanges 512 | jq '.[]' | sed -n '257p'
"8000-8080"
```

mattlord commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+// Normalize adds padding to id as necessary to make sure it can be compared as an 8 byte integer.
+// From a key range perspective, -80 == 00-80 == 0000-8000 == 000000-800000, etc. and they should
+// always be treated the same even if they are different lengths.
+func Normalize(id []byte) []byte {
+	if len(id) >= 8 {
+		return id[:8]
```
I'm not sure that this is expected or correct behavior. 

Wouldn't it be better to do this?
```
func Normalize(id []byte) ([]byte, error) {
            if !IsValidKey(id) {
                return fmt.Errorf("provided keyspace ID %q is outside of the valid unsigned 64 bit range", hex.EncodeToString(id))
```

We'd then have to thread that back through compare. This ties into my error handling comment in `TestCompare`.

What do you think?

mattlord commented on go/vt/key/key_test.go:
```diff
@@ -460,7 +685,62 @@ func TestParseShardingSpec(t *testing.T) {
+		args  args
+		wants wants
+	}{
+		{"a and b are both full range", args{stringToKeyRange("-"), stringToKeyRange("-")}, wants{0, true, 0, true, 0, true}},
```
This would be much more readable and easier to confirm expected results:
```
{
    name: "a and b are both full range",
    args: args{
       stringToKeyRange("-"), stringToKeyRange("-")
    },
    wants: wants{
       wantStartCompare: 0,
       wantStartEqual: true,
       wantEndCompare: 0,
       wantEndEqual: true,
       wantCompare: 0,
       wantEqual: true,
    }
},
```

jeremycole replied:
You just need a good IDE, Matt! But yes, that's fine to me, I'm looking at it always with the fields named by the IDE, so it's no difference to me and makes it easier to see what's going on in other IDEs and on e.g. Github.

mattlord reviewed (commented):
This is great! Nice work on this, @jeremycole !  ❤️ I only had some minor nits/comments/suggestions. Please let me know what you think.

deepthi commented:
@jeremycole others have done a detailed review, so I just wanted to chime in say that this is great work! The other proposed changes all seem to be good ones from a code readability / quality / maintainability point of view as well, so there is no objection to proceeding along those lines.

You did say in the description that you ran into a bug. Is there an issue for that which can be linked here? If not, do you mind creating one?
Will also be good to create an issue for the proposed improvements to link to an eventual PR.

harshit-gangal commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
-	}
-	if n&(n-1) != 0 {
-		return nil, fmt.Errorf("the shard count must be a power of two: %v", n)
+// Normalize adds padding to id as necessary to make sure it can be compared as an 8 byte integer.
+// From a key range perspective, -80 == 00-80 == 0000-8000 == 000000-800000, etc. and they should
+// always be treated the same even if they are different lengths.
```
I also thought about it like `-ff` but those are just weird cases, Definitely, they can be handled though.

jeremycole replied:
Due to the recent changes to remove trailing zeroes rather than add padding, the `0` case and `00-` case are handled automatically (munged into the same thing as `-`). I don't think there's anything special about `-ff` at all (it's just a regular value).

harshit-gangal commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+// Normalize adds padding to id as necessary to make sure it can be compared as an 8 byte integer.
+// From a key range perspective, -80 == 00-80 == 0000-8000 == 000000-800000, etc. and they should
+// always be treated the same even if they are different lengths.
+func Normalize(id []byte) []byte {
+	if len(id) >= 8 {
+		return id[:8]
```
One breaking change that we should talk about is restricting it is 8-byte length. which the current implementation does not do.
Current implementation support keyspace id to be arbitrary length based on the Vindex.

deepthi replied:
An example of longer-than-8-byte keyspace IDs can be seen [here](https://vitess.io/docs/16.0/concepts/shard/#shard-naming) (link courtesy @harshit-gangal)

mattlord replied:
Yeah, I (also) mistakenly thought they were always 64 bits, but that's only the length of the default `hash` vindex based keyspace IDs. Some are shorter and others longer. You can see examples of how many hex digits are produced for each vindex type by using the vindex query functions: https://vitess.io/docs/17.0/reference/features/vindexes/#query-vindex-functions 

I was also pointed this blog post that I don't remember previously reading: https://vitess.io/blog/2017-09-18-custom-sharding-with-vitess/

jeremycole replied:
I see what you mean here. I am not sure that's a real case that anyone is using but I am quite okay with not changing the semantics here. We can pivot the `Normalize` method to, instead of padding out to a fixed length (which there isn't any value that would satisfy the requirements without defining a maximum length) to _instead_ remove all trailing zeroes consistently.

I made that change locally and it works fine and all tests pass the same except for `TestNormalize` (since it's explicitly checking the return values of `Normalize`, so it would be expected). I'll push another commit for this.

jeremycole commented on go/vt/key/key.go:
```diff
@@ -26,16 +26,19 @@ import (
 
+var (
+	ZeroKey         = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
+	KeyRangePattern = regexp.MustCompile(`^[0-9a-fA-F]*-[0-9a-fA-F]*$`)
```
I agree, I only relocated the existing regexp constant from where it was placed far below in the same file but it does need work; in fact it also needs to only accept multiples of two digits. I would propose leaving fixing this up to a follow-up PR though, so that the actual fix and associated tests can be much more clear. I'll be happy to create an issue for that and even work on it though.

mattlord replied:
Resolving this since now I know it's of variable length.

mattlord commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
-	}
-	return &topodatapb.KeyRange{Start: startBytes, End: endBytes}, nil
+// Compare compares two keyspace IDs while taking care to normalize them; returns -1 if a<b, 1 if a>b, 0 if equal.
+func Compare(a, b []byte) int {
+	return bytes.Compare(Normalize(a), Normalize(b))
+}
+
+// Less returns true if a is less than b.
+func Less(a, b []byte) bool {
+	return Compare(a, b) < 0
+}
+
+// Equal returns true if a is equal to b.
+func Equal(a, b []byte) bool {
+	return Compare(a, b) == 0
 }
```
Resolving this since I now realize they are NOT 64 bit and keyspace IDs are an arbitrary number of bytes.

jeremycole commented:
Thanks for the prompting, I filed https://github.com/vitessio/vitess/issues/12535 with the actual bug we hit that caused us to go down this rabbit hole. :)

jeremycole commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+
+// KeyRangeIsComplete returns true if the KeyRange covers the entire keyspace.
+func KeyRangeIsComplete(keyRange *topodatapb.KeyRange) bool {
+	return keyRange == nil || (Empty(keyRange.Start) && Empty(keyRange.End))
```
This is a good point, we should handle an `id` with any amount of zero bytes actually, and I think can be solved with a very small change: Empty should actually check `len(Normalize(id))` rather than `len(id)` – with my latest change to remove trailing zeroes in `Normalize()`, the length of an all-zero `id` would always be zero as well.

jeremycole commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
+// KeyRangesIntersect returns true if some part of KeyRange a and b overlap, meaning that some keyspace ID values
+// exist in both a and b.
+func KeyRangesIntersect(a, b *topodatapb.KeyRange) bool {
+	if a == nil || b == nil {
```
Yep, I think this is a good change.

jeremycole commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
-func KeyRangesIntersect(first, second *topodatapb.KeyRange) bool {
-	if first == nil || second == nil {
-		return true
+// KeyRangesOverlap returns a KeyRange representing the overlap between KeyRange a and b. If they do not overlap,
+// an error is returned.
+func KeyRangesOverlap(a, b *topodatapb.KeyRange) (*topodatapb.KeyRange, error) {
```
I did not implement this originally, and I didn't change its interface... but actually, @hkdsun points out that `GetKeyRangeOverlap()` is actually not used in the codebase _except for its own tests_, so I would propose we actually just delete `KeyRangesOverlap` and the new supporting `keyRangeStartMax` and `keyRangeEndMin` functions. Any reason y'all can think of we should not do that?

jeremycole commented on go/vt/key/key.go:
```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
-// KeyRangeIncludes returns true if the first provided KeyRange, big,
-// contains the second KeyRange, small. If they intersect, but small
-// spills out, this returns false.
+// KeyRangeIncludes returns true if KeyRange big fully contains KeyRange small.
 func KeyRangeIncludes(big, small *topodatapb.KeyRange) bool {
```
Currently that function name is already used for `KeyRangeContains(keyRange *topodatapb.KeyRange, id []byte)` and Go doesn't support method overloading. It could be named `KeyRangeContainsKeyRange (a, b *topodatapb.KeyRange)` though! What do you think?

jeremycole commented on go/vt/key/key.go:
```diff
@@ -351,11 +351,63 @@ func ParseShardingSpec(spec string) ([]*topodatapb.KeyRange, error) {
+// We do this to produce the same KeyRange objects as ParseKeyRangeParts() does.
+// Because it's using the Go hex methods, it's omitting trailing zero bytes as
+// well.
+func EvenShardsKeyRange(i, n int) (*topodatapb.KeyRange, error) {
+	if n <= 0 {
+		return nil, fmt.Errorf("the shard count must be > 0: %v", n)
+	}
+	if i >= n {
+		return nil, fmt.Errorf("the index of the shard must be less than the total number of shards: %v < %v", i, n)
+	}
+	if n&(n-1) != 0 {
+		return nil, fmt.Errorf("the shard count must be a power of two: %v", n)
+	}
```
As I noted in the PR description, I would actually propose deleting `EvenShardsKeyRange` and `GenerateShardRanges` in a follow-up PR. I just moved them here while reorganizing the file a bit.

jeremycole commented:
👋 I think I have addressed all review feedback with the last commits pushed. Sorry for adding everyone as reviewers, every time I touch a new file it seems we get a new reviewer added automatically. 😄 

I broke up the additional work into several separate commits so that those changes can more easily be reviewed in isolation if you so desire. Let me know anything else you need! (And we'll see how the full CI runs go.)

deepthi commented:
Looks like you broke something, both `unit_test` and local example are failing. On the plus side, they should be easy to reproduce locally!

jeremycole commented:
The failure in `unit_test` was a flake unrelated to my changes, seems to be a timeout or something in gRPC stuff resulting in `FAIL: TestMtlsAuth (60.11s)`. It succeeds locally.

I haven't quite gotten the `local_example` to work locally but in the mean time would y'all mind rerunning those tests?

deepthi commented:
Looks like the local example failure is a flake as well. I see it failing on a PR that touches no code here https://github.com/vitessio/vitess/pull/12566
https://github.com/vitessio/vitess/actions/runs/4356176371/jobs/7613821315

deepthi commented:
@jeremycole I re-ran the 3 failing tests. I've also invited you to the org, so that in future you will have the ability to this too. Invite goes to your primary GH email.

jeremycole commented:
@mattlord does this address your comments and concerns?

jeremycole commented:
👋 We're still holding on this but using it in our internal branch. Are there any remaining concerns about it?

mattlord commented:
> 👋 We're still holding on this but using it in our internal branch. Are there any remaining concerns about it?

Hey! Sorry for the delay. I'll add this review to my ToDo for tomorrow.

mattlord reviewed (approved):
Nice work on this, @jeremycole ! The quality of the code, comments, PR description, issue etc are all very good. ❤️
